keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1365 @@
+
+If the parent object is cluster-scoped, member objects by default are at the cluster scope. The `applyset.k8s.io/additional-namespaces` annotation can be used to allow member objects in one or more namespaces.
+
+### Tooling Interoperability
+
+There is a rich ecosystem of existing tooling that we hope will adopt these labels and annotations.  So that different tooling can interoperate smoothly, we define some requirements for safe interoperability here.
+
+For read operations, we expect that using different tooling shall generally be safe.  As these labels do not collide with existing tooling, we would expect that objects installed with existing tooling would be invisible to the porcelain tooling until they had been updated to include the labels.  We do not propose to implement “bridges” to existing tooling, rather as the proposal here is lightweight and small, it makes more sense to update the existing tooling.  We may add warnings such as “ApplySets using an old version of X detected, upgrade to v123 of X to work with those ApplySets”.
+
+For write operations, we need to be more careful.  Deleting an ApplySet using the “wrong tool” should be safe, but we will likely include a confirmation if deleting an ApplySet using the “wrong tool”, particularly unknown tools.  We expect that porcelain tools may define richer behavior on delete, so this is the equivalent of pulling the power cable on an ApplySet instead of performing a clean shutdown.
+
+We do not believe that update operations are safe if using the “wrong tool”, because that tooling may have additional metadata that would then not be updated.  Tooling should generally reject applying on top of unknown ApplySets.  Porcelain tooling may choose to recognize other tooling and implement specific logic there; in particular this may be useful for moving between different major versions of the same tooling.
+
+In order to identify usage of the "wrong tool", we rely on the `applyset.k8s.io/tooling` annotation,
+which tooling can set to protect their ApplySets.
+Specification-compliant porcelain tooling MUST recognize that
+a different tool is managing the ApplySet and provide an appropriate error or warning.
+We intend to explore the trade-off between safety and user-friendly behaviour
+here, during evolution of the feature in alpha and beyond.
+
+### Objects with owner references
+
+If an object in the set retrieved for pruning has owner references,
+tooling should verify that those references match the ApplySet parent.
+If they do, the tool should proceed as normal. If they do not, the
+tooling should consider this an ownership conflict and throw an error.
+
+We are taking this stance initially to be conservative and ensure that
+use cases related to objects bearing owner references are surfaced.
+In the future, we could downgrade this stance to recommending a warning,
+or to considering owner references orthogonal and ignoring them entirely.
+
+### Versioning
```

Comment:
This section is another argument for using fields in CRDs :)

Reply:
I'd love to suggest a CRD, but I'm not naive enough to believe that my CRD will fare any better than those that came before.  I think though that the pragmatic answer is that we enable usage of CRDs, but we don't rely on cluster administrators installing one.  If one happens to come along that delivers exactly the fields we need, that would indeed be wonderful.

Maybe we could have namespace scoped CRDs?

Reply:
Why are we limiting the discussion to CRDs? Why not built-in types? This is a plumbing level feature. If we have a built-in applyset + object ref to tooling porcelain object, we would avoid the adoption issue related to needing to install the CRD, and have better discovery, and have better options for validation and consistency management of the objects.